x-pack/plugins/observability_ai_assistant/public/utils/get_timeline_items_from_conversation.tsx

```diff
@@ -73,50 +95,59 @@ export function getTimelineItemsfromConversation({
           // User executed a function:
-          if (message.message.name && functionCall) {
-            title = i18n.translate('xpack.observabilityAiAssistant.executedFunctionEvent', {
-              defaultMessage: 'executed the function {functionName}',
-              values: {
-                functionName: message.message.name,
-              },
-            });
-
-            content = convertFunctionParamsToMarkdownCodeBlock({
-              name: message.message.name,
-              arguments: JSON.parse(functionCall.arguments || '{}'),
-            });
 
-            element = chatService.hasRenderFunction(message.message.name) ? (
-              <RenderFunction
-                name={message.message.name}
-                arguments={functionCall?.arguments}
-                response={message.message}
-              />
-            ) : null;
+          if (message.message.name && functionCall) {
+            const parsedContent = JSON.parse(message.message.content ?? 'null');
+            const isError = !!(parsedContent && 'error' in parsedContent);
+
+            title = !isError
+              ? i18n.translate('xpack.observabilityAiAssistant.executedFunctionEvent', {
+                  defaultMessage: 'executed the function {functionName}',
+                  values: {
+                    functionName: message.message.name,
+                  },
+                })
+              : i18n.translate('xpack.observabilityAiAssistant.executedFunctionFailureEvent', {
+                  defaultMessage: 'failed to execute the function {functionName}',
+                  values: {
+                    functionName: message.message.name,
+                  },
+                });
+
+            element =
+              !isError && chatService.hasRenderFunction(message.message.name) ? (
+                <RenderFunction
+                  name={message.message.name}
+                  arguments={functionCall?.arguments}
+                  response={message.message}
+                />
+              ) : undefined;
+
+            content = !element ? convertMessageToMarkdownCodeBlock(message.message) : undefined;
 
-            canCopy = true;
-            canEdit = hasConnector;
-            canGiveFeedback = true;
-            canRegenerate = hasConnector;
-            collapsed = !Boolean(element);
-            hide = false;
+            canEdit = false;
+            collapsed = !isError && !element;
           } else {
             // is a prompt by the user
             title = '';
             content = message.message.content;
 
-            canCopy = true;
             canEdit = hasConnector;
-            canGiveFeedback = false;
-            canRegenerate = false;
             collapsed = false;
-            hide = false;
           }
 
           break;
 
         case MessageRole.Assistant:
+          canRegenerate = hasConnector;
+          canCopy = true;
+          canGiveFeedback = true;
+          hide = false;
```

Comment:
you can always regenerate a response from the LLM (and give feedback). Other than that, simply moving some common assignments to the top.